concordium-node/src/bin/cli.rs

```diff
@@ -129,30 +108,99 @@ async fn main() -> anyhow::Result<()> {
     )?;
     info!("Consensus layer started");
 
-    // A flag to record that the import was stopped by a signal handler.
+    // Setup task with signal handling before doing any irreversible operations
+    // to avoid being interrupted in the middle of sensitive operations, e.g.,
+    // creating the database.
+    let (mut shutdown_receiver, shutdown_sender) = setup_shutdown_signal_handling();
+
+    // A flag to record that out-of-band-catchup import was stopped by a signal
+    // handler.
     let import_stopped = Arc::new(atomic::AtomicBool::new(false));
+    // Flag to record errors in the services running in the node.
+    let rpc_error = Arc::new(atomic::AtomicBool::new(false));
+    let grpc2_error = Arc::new(atomic::AtomicBool::new(false));
+    // Set up handler for setting the flags and terminating the block state import.
     {
-        // set up the handler for terminating block state import.
         let consensus = consensus.clone();
         let import_stopped = Arc::clone(&import_stopped);
+        let rpc_error = Arc::clone(&rpc_error);
+        let grpc2_error = Arc::clone(&grpc2_error);
+        let mut shutdown_receiver = shutdown_sender.subscribe();
         tokio::spawn(async move {
-            if shutdown_signal_2.recv().await.is_err() {
-                error!("Signal handler dropped. This should not happen.");
+            // Set flags indicating if a service crashed. This thread contains a loop since
+            // errors in services may occur after a node shutdown is initiated.
+            loop {
+                match shutdown_receiver.recv().await {
+                    // The RPC server crashed.
+                    Ok(NodeShutdownCause::RPCServer) => {
+                        rpc_error.store(true, atomic::Ordering::Release);
+                    }
+                    // The GRPC2 server crashed.
+                    Ok(NodeShutdownCause::GRPC2Server) => {
+                        grpc2_error.store(true, atomic::Ordering::Release);
+                    }
+                    // The statistics service crashed.
+                    Ok(NodeShutdownCause::StatsServer) => {}
+                    // A signal was sent to terminate the node.
+                    Ok(NodeShutdownCause::Signal) => {}
+                    Err(e) => {
+                        error!(
+                            "Node shutdown channel was dropped unexpectedly. This should not \
+                             happen: {}",
+                            e
+                        );
+                        break;
+                    }
+                }
+                // If we are here, either an error occurred or a signal to shut down the node
+                // was made by the user. In any case, we stop the out-of-band-catchup.
+                match import_stopped.compare_exchange(
+                    false,
+                    true,
+                    atomic::Ordering::Acquire,
```

Comment:
I will probably have to look into memory ordering options to see if these options are sensible.